### PR TITLE
Support extra elements in pattern matching

### LINKED ISSUE #238: allow extra elements in pattern matching
### Use-cases
As mentioned in #132 and #190, pattern matching as below needs to be supported:
- fix `let = [1, 2]; x` (should fail)
- `let [x, y, ...] = [1, 2, 3]; x + y`  Should pass.
- `let [_, ...t] = [1, 2, 3, 4, 5]; t sum . ` Should pass.
- `[a, b, c, ...t]` allows extra elements in addition to the explicitly bounds ones and binds `t` to an array of them. (Maybe...)
- `(m: x, n: y, ...t)` allows extra elements in addition to the explicitly bounds ones and binds `t` to a tuple of them.
- `{a: x, b: y, ...t}` allows extra elements in addition to the explicitly bounds ones and binds `t` to a dict of them.
- `[a, ...t, b, c]`. The `...t` part can appear anywhere. (Maybe...)
- `let (:lower, :upper, ...) = //str; 1`
- `[a, b, ...t, c, d, ...u, e, f]`. Multiple `...` parts may lead to non-deterministic bindings, depending on the constraints imposed by patterns, a&ndash;f. (Maybe not...)

### References
- #132

### PR DESCRIPTION
**Pull request by ChloePlanet**
Fixes #238 .

Changes proposed in this pull request:
As mentioned in #132 and #190, pattern matching as below needs to be supported:
- [x] fix `let [x] = [1, 2]; x` (should fail)
- [x] `let [x, y, ...] = [1, 2, 3]; x + y`  Should pass.
- [x] `let [_, ...t] = [1, 2, 3, 4, 5]; t sum . ` Should pass.
- [x] `[a, b, c, ...t]` allows extra elements in addition to the explicitly bounds ones and binds `t` to an array of them.
- [x] `(m: x, n: y, ...t)` allows extra elements in addition to the explicitly bounds ones and binds `t` to a tuple of them.
- [x] `{a: x, b: y, ...t}` allows extra elements in addition to the explicitly bounds ones and binds `t` to a dict of them.
- [x] `[a, ...t, b, c]`. The `...t` part can appear anywhere.
- [x] `...` should support an empty tail. Likewise, all uses of `...` should support an empty residue.

Checklist:
- [x] Added related tests

### CODE REVIEW

File: syntax/expr_let_test.go [hunk 1]
```diff
@@ -49,22 +49,26 @@ func TestExprLetArrayPattern(t *testing.T) {
...
-	AssertCodeErrors(t, `let (a:x, a:x) = (a:4); x`, "")
-	AssertCodeErrors(t, `let x = 5; let (a:(x)) = (a:4); x`, "")
-	AssertCodeErrors(t, `let (a:x, b:x) = (a:4, b:7); x`, "")
-	AssertCodeErrors(t, `let x = 5; let (a:[(x)]) = (a:[4]); x`, "")
+	AssertCodeErrors(t, `let (a: x) = (b: 7, a: 4); x`, "")
+	AssertCodeErrors(t, `let (a: x, a: x) = (a: 4, a: 4); x`, "")
+	AssertCodeErrors(t, `let (a: x, a: x) = (a: 4); x`, "")
+	AssertCodeErrors(t, `let x = 5; let (a: (x)) = (a: 4); x`, "")
+	AssertCodeErrors(t, `let (a: x, b: x) = (a: 4, b: 7); x`, "")
+	AssertCodeErrors(t, `let x = 5; let (a: [(x)]) = (a: [4]); x`, "")
```
Comment (marcelocantos) [inline syntax/expr_let_test.go:64]: :ok_hand:

File: syntax/expr_let_test.go [hunk 2]
```diff
@@ -49,22 +49,26 @@ func TestExprLetArrayPattern(t *testing.T) {
...
 }
 
 func TestExprLetDictPattern(t *testing.T) {
+	AssertCodesEvalToSameValue(t, `42`, `let {1: a} = {1: 42}; a`)
+	AssertCodesEvalToSameValue(t, `42`, `let {[1, 2, 3]: a} = {[1, 2, 3]: 42}; a`)
+	AssertCodesEvalToSameValue(t, `42`, `let a = 4; let {"x": a} = {"x": 42}; a`)
+	AssertCodesEvalToSameValue(t, `42`, `let a = 42; let {"x": (a)} = {"x": 42}; a`)
```
Comment (marcelocantos) [inline syntax/expr_let_test.go:71]: nit: One or two cases that combine `ident` and `(ident)` would be good. Consider the same for other pattern types.

File: syntax/expr_let_test.go [hunk 3]
```diff
@@ -74,3 +78,25 @@ func TestExprLetDictPattern(t *testing.T) {
...
+	AssertCodesEvalToSameValue(t, `[2, [3, 4, 5]]`, `let [_, x, ...t] = [1, 2, 3, 4, 5]; [x, t]`)
+	AssertCodesEvalToSameValue(t, `[1, 4, 5, [2, 3]]`, `let [x, ...t, y, z] = [1, 2, 3, 4, 5]; [x, y, z, t]`)
+	AssertCodesEvalToSameValue(t, `[4, 5, [2, 3]]`, `let [_, ...t, y, z] = [1, 2, 3, 4, 5]; [y, z, t]`)
+	AssertCodesEvalToSameValue(t, `[1, 2]`, `let (m: x, n: y, ...) = (m: 1, n: 2, j: 3, k: 4); [x, y]`)
+	AssertCodesEvalToSameValue(t, `[1, 2, (j: 3, k: 4)]`, `let (m: x, n: y, ...t) = (m: 1, n: 2, j: 3, k: 4); [x, y, t]`)
+	AssertCodesEvalToSameValue(t, `[1, 2]`, `let {"m": x, "n": y, ...} = {"m": 1, "n": 2, "j": 3, "k": 4}; [x, y]`)
+	AssertCodesEvalToSameValue(t, `[1, 2, {"j": 3, "k": 4}]`,
+		`let {"m": x, "n": y, ...t} = {"m": 1, "n": 2, "j": 3, "k": 4}; [x, y, t]`,
+	)
+	AssertCodeErrors(t, `let [x, y, ...] = [1, 2]; x`, "")
```
Comment (marcelocantos) [inline syntax/expr_let_test.go:98]: `...` should support an empty tail. Likewise, all uses of `...` should support an empty residue.

File: syntax/expr_let_test.go (hunk 2 above)
Comment (ChloePlanet) [reply syntax/expr_let_test.go:71]: Do you mean `let {"x": a, "y": (a)} = {"x": 42, "y": 42}; a`?

Comment (marcelocantos) [reply syntax/expr_let_test.go:71]: Yep, maybe also `b` and `(a)`.

Comment (marcelocantos) [reply syntax/expr_let_test.go:71]: Oh, actually, also `let {"x": a, "y": (a)} = {"x": 42, "y": 56}; a` (should equal 42). The `a` is a new binding, while the `(a)` is an existing binding. They should have nothing to do with each other.

Comment (ChloePlanet) [reply syntax/expr_let_test.go:71]: `let a = 56; let {"x": a, "y": (a)} = {"x": 42, "y": 56}; a` should equal 42?

**Pull request merged**